### PR TITLE
ADNI Classifier based on a Siamese Neural Network

### PR DESCRIPTION
**Pull request by Yoxiourn**
Implemented a Classifier based on a Siamese Neural Network using Tensorflow. The dataset used was the ADNI dataset, and the data is pooled together before being split into 60/20/20 training, validation and testing proportions. The Siamese Neural Network is very standard, with 2D Convolutional and Max Pooling layers. The classifier is based on the embedding network of the Siamese Neural Network (that is, the network each of the twins uses for transforming an image into a vector) and the distance computation used is L1 distance. The best model architecture produced an SNN testing accuracy of 86.12%, and the classifier based on the embedding network of the SNN reached an accuracy on the entire dataset of 88.79%. Although the performance of the architecture does seem to fluctuate quite significantly between runs, it does generally reach more than 70% accuracy for both the SNN and the classifier.

### CODE REVIEW

**Comment by Yoxiourn** [comment]
Also just in case you need:

Name: Matthew Shearer
Student Number: 4642283

**Comment by nathasha-naranpanawa** [comment]
- Commit messages are OK
- Readme is OK. Examples of input/output images not discussed
- Missing information on preprocessing the data on Readme. Was the data used for training 2D or 3D? (Looking at the code they might be 2D)
- Plots for train/validation loss and accuracy shown
- Performance is OK-ish. Although training loss and accuracy are fine, validation loss and accuracy fluctuates a lot, and attempts to fix that have not been specifically mentioned
- Looking at the code, the above error could be due to the fact that the loss function used in the Siamese is not contrastive loss but a classification loss (slightly incorrect)
- No prediction results shown with example images

**Comment by Yoxiourn** [comment]
Hi Nathasha,

Thanks for the feedback! I was using 2D for training (I apologize for not making this clear in the readme). I was not aware of that incorrect loss function, so thank you for pointing that out! For the validation loss and accuracy fluctuations, I haven't specifically addressed how I've tried to combat this, but the true answer is that I just tried a bunch of different architectures (including different epochs, loss functions and CNNs, with some of these committed), and I found that ones that had little fluctuations tended to converge to a low accuracy (~60%) while other architectures (like this) could fluctuate a lot more but reach higher validation and testing accuracies (~70% to ~80%).

Sorry for not showing prediction results.

I assume that I'm not supposed to edit anything from this feedback yet (looking at other messages from other tutors) so I will leave this as is for now.

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation (some loss fluctuation, but explained in comment) -1
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage (no predictions) -1
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable (discussion would be good) -1
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
